lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+      // iterating over every single return instruction in the function and
+      // its type. This will get confused if there are multiple return
+      // instructions. I haven't gotten this to happen from C yet but it is
+      // to construct a trivial example where this breaks in bitcode. I think
+      // its possible that I only need to consider the first return instruction
+      // that I see, because they all should be of the same type but I am not
+      // sure.
+      if (auto return_inst = llvm::dyn_cast<llvm::ReturnInst>(&inst)) {
+        anvill::ValueDecl value_declaration = {};
+
+        const llvm::Value *value = return_inst->getReturnValue();
+        if (!value) continue;
+        value_declaration.type = value->getType();
+
+        switch (value_declaration.type->getTypeID()) {
+          case llvm::Type::IntegerTyID:
+          case llvm::Type::PointerTyID: {
+            // Allocate RAX for an integer or pointer
+            value_declaration.reg =
+                new remill::Register("RAX", 0, 8, 0, value_declaration.type);
+            break;
+          }
+          case llvm::Type::FloatTyID: {
+            // Allocate XMM0 for a floating point value
+            value_declaration.reg =
+                new remill::Register("XMM0", 0, 16, 0, value_declaration.type);
+            break;
+          }
+          case llvm::Type::StructTyID: {
+            // Try to split the struct over the registers
+            std::vector<bool> allocated(return_register_constraints.size(),
+                                        false);
+            auto struct_ptr =
+                llvm::cast<llvm::StructType>(value_declaration.type);
+            auto mapping = TryReturnThroughRegisters(
+                *struct_ptr, return_register_constraints);
+            if (mapping) {
+              // There is a valid split over registers, so add the mapping
+              return *mapping;
+            } else {
+              // Struct splitting didn't work so do RVO. Assume that the pointer
+              // to the return value resides in RAX.
+              value_declaration.reg =
+                  new remill::Register("RAX", 0, 8, 0, value_declaration.type);
+            }
+            break;
+          }
+          default: {
+            LOG(ERROR) << "Encountered an unknown return type";
+            exit(1);
+          }
+        }
+
+        return_value_declarations.push_back(value_declaration);
+      }
+    }
+  }
+
+  return return_value_declarations;
+}
+
+std::vector<anvill::ParameterDecl> X86_64_SysV::BindParameters(
+    const llvm::Function &function) {
+  std::vector<anvill::ParameterDecl> parameter_declarations;
+  auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Used to keep track of which registers have been allocated
+  std::vector<bool> allocated(parameter_register_constraints.size(), false);
+
+  // Stack position of the first argument
```

Comment:
Larger comment describing stack spilling behavior of arguments. Perhaps some small ASCII diagrams or just tabular layout so that it's documented in what order arguments passed on the stack are pushed on the stack.